[Build] Update Kotlin to 1.9.25 (++)

ParaskP7 opened the pull request:
Depends On:
 - [x] [ADC#34](https://github.com/Automattic/android-dependency-catalog/pull/34)
 - [x] Update ADC to `1.18.0`

### Description

This PR updates Kotlin to [1.9.25](https://github.com/JetBrains/kotlin/releases/tag/v1.9.25).

This update isn't that necessary, but should be considered anyway as it will do the following:
- Align this library, to that Kotlin version, which is already used by most of this library's clients and libraries.
- Unblock the usage of the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin for this project.
- Prepares this project for the Kotlin 2+ major update.

FYI: Since updating Kotlin alone wasn't enough, due to various build failures, this update also incorporates the below 4 additional updates:
- `detekt` -> [1.23.6](https://github.com/detekt/detekt/releases/tag/v1.23.6)
- `androidx-room` -> [2.6.1](https://developer.android.com/jetpack/androidx/releases/room#2.6.1)
- `dagger` -> [2.51.1](https://github.com/google/dagger/releases/tag/dagger-2.51)
- `kotlinx-coroutines` -> [1.8.1](https://github.com/Kotlin/kotlinx.coroutines/releases/tag/1.8.1)

### Test

- Smoke test the `FluxC Example` app, try out all available screens and functionality. Verify everything is working as expected.
- In addition to that, please be thorough about reviewing this `Kotlin` and 4 extra updates, by quickly smoke test both, the [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android) and [WCAndroid](https://github.com/woocommerce/woocommerce-android), with this version of `FluxC`, or via composite builds, and see if everything is working as expected.

ParaskP7 committed:
Deps: Update to kotlin 1.9.25 (via catalog's pr-sha combination)

FYI: This change was done for testing purposes and until the below
'Android Dependency Catalog' #34 PR gets merged to 'trunk'. When that's
done, the 'catalogVersion' will be updated to '1.18.0' instead.

------------------------------------------------------------------------

Since updating Kotlin isn't enough by itself, due to various build
failures, this update also incorporates the below 4 updates:
- 'detekt' -> '1.23.6'
- 'androidx-room' -> '2.6.1'
- 'dagger' -> '2.51.1'
- 'kotlinx-coroutines' -> '1.8.1'

Android Dependency Catalog PR: [Dependency Updates] Update Kotlin to
1.9.25 #34
- https://github.com/Automattic/android-dependency-catalog/pull/34

ParaskP7 committed:
Build: Add jvm target to point to java 8 to kotlin options

This change is necessary because otherwise the build fails with the
below exception:

------------------------------------------------------------------------

> Task :fluxc-processor:kaptGenerateStubsKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':fluxc-processor:kaptGenerateStubsKotlin'.
> Inconsistent JVM-target compatibility detected for tasks 'compileJava'
 (1.7) and 'kaptGenerateStubsKotlin' (17).

  Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
  Learn more about JVM-target validation: https://kotl.in/gradle/jvm/
   target-validation

ParaskP7 committed:
Build: Move java related source/target compatibility into java block

This solves the 'Assignment is not used' build warnings on the
'fluxc-annotations' and 'fluxc-processor' Java related modules.

ParaskP7 committed:
Fix: Resolve coroutine related send blocking unresolved reference

Error Message: "Unresolved reference: sendBlocking"

ParaskP7 committed:
Fix: Resolve coroutine related offer error

Error Message: "Using 'offer(E): Boolean' is an error. Deprecated in the
favour of 'trySend' method"

ParaskP7 committed:
Fix: Add any as upper bound for list item to make it non-nullable

Error Messages:
- "'onItemAtEndLoaded' overrides nothing"
- "Type mismatch: inferred type is LIST_ITEM but LIST_ITEM
& Any was expected"

ParaskP7 committed:
Fix: Remove job from flow with default context to make it safe for flows

Otherwise a runtime exception is thrown whenever
'flowWithDefaultContext(...)' is being used, tests included
(see 'OrderUpdateStoreTest').

------------------------------------------------------------------------

java.lang.IllegalArgumentException:
 Flow context cannot contain job in it. Had [
  kotlinx.coroutines.test.TestCoroutineScheduler@53a665ad,
  kotlinx.coroutines.test.TestCoroutineExceptionHandler@2c0b4c83,
  JobImpl{Active}@abff8b7,
  TestCoroutineDispatcher[
   scheduler=kotlinx.coroutines.test.TestCoroutineScheduler@53a665ad
  ]
 ]

ParaskP7 committed:
Fix: Resolve room related cascade/replace unresovled references

Error Messages:
- "Unresolved reference: CASCADE"
- "Unresolved reference: REPLACE"

ParaskP7 committed:
Fix: Resolve kotlin related exhaustive when expression errors

Error Message: " 'when' expression must be exhaustive, add necessary
'XYZ' branch or 'else' branch instead"

ParaskP7 committed:
Database: Update latest database schema for both wp/wc android databases

ParaskP7 committed:
Analysis: Remove deprecated potential bugs rule from detekt's config

Otherwise running 'detekt' results in a build failure:

------------------------------------------------------------------------

> Task :example:detekt FAILED
Property 'potential-bugs>MissingWhenCase' is deprecated. Rule deprecated
 as compiler performs this check by default.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':example:detekt'.
> Run failed with 1 invalid config property.
        - Property 'potential-bugs>MissingWhenCase' is deprecated. Rule
         deprecated as compiler performs this check by default.

ParaskP7 committed:
Analysis: Resolve unused private property warning on woo admin fragment

Warning Message: "Private property `gson` is unused."

ParaskP7 committed:
Analysis: Resolve for each on range warning on view group ext

Warning Message: "Using the forEach method on ranges has a heavy
performance cost. Prefer using simple for loops."

ParaskP7 committed:
Analysis: Resolve wildcard import warnings on custom redirect int. test

Warning Messages:
- "okhttp3.* is a wildcard import. Replace it with fully qualified
imports."
- "org.junit.Assert.* is a wildcard import. Replace it with fully
qualified imports."

ParaskP7 committed:
Analysis: Resolve unused parameter warning on coupons dao test

Warning Message: "Function parameter `id` is unused."

ParaskP7 committed:
Analysis: Resolve use require warnings for plugins woocommerce module

Warning Message: "Use require() instead of throwing an
IllegalArgumentException."

wzieba commented on fluxc/schemas/org.wordpress.android.fluxc.persistence.WPAndroidDatabase/27.json:
Why are these updates necessary? And why in this specific migration files? Looking at e.g. `26.json`, the `autoGenerate` key is below `columnNames`

ParaskP7 replied:
This was done automatically (as usual), I just committed the changes. I guess, most probably due to the `room` update ([2.6.1](https://developer.android.com/jetpack/androidx/releases/room#2.6.1)), the structure of those `*.json` files is now updated too... 🤷

wzieba replied:
> I guess, most probably due to the room update ([2.6.1](https://developer.android.com/jetpack/androidx/releases/room#2.6.1)), the structure of those *.json files is now updated too... 🤷

But in this case, I'd expect all other `*.json` file to be updated as well, not only very specific `27.json`.

I've reverted the ordering in this file and re-run build, and it wasn't changed back.

I'm not very concerned about this particular change, as it's only ordering, but more about a precedence of editing old database schema, which should never be edited. Taking all of this into account, can we revert this change?

ParaskP7 replied:
> But in this case, I'd expect all other *.json file to be updated as well, not only very specific 27.json.

Well, I am not sure what I would expect to be honest... 🤷 

> I've reverted the ordering in this file and re-run build, and it wasn't changed back.

Yea, I tried that myself just now too, after reverting this change, and couldn't make this change back... 🤷 

PS: This behavior was consistent for me, because it happened twice to me, when I was investigating how to update Kotlin to `1.9.25`, and then, once more, when I actually applied my solution (post investigation). So, I personally saw that happening automatically, and thus, wanted to commit that change so that other won't have to afterwards, thinking that if I don't commit it now, it will anyway pop-up later for someone else. 🤔 

> I'm not very concerned about this particular change, as it's only ordering...

Exactly my thoughts! 👍 

> ...but more about a precedence of editing old database schema, which should never be edited.

This is a bit debatable, but I understand the concern and I find it valid nevertheless. 👍 

FYI: When I did some research in the past on these updates, I figured out that those are not mandatory and will not fail the migration and as such upgrading the app. It is just a good practice to have this file committed in the version control system and have it in the codebase. For more info check the [documentation](https://developer.android.com/reference/android/arch/persistence/room/Database#exportSchema()).

> Taking all of this into account, can we revert this change?

Done: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3059/commits/ce00973b295ca8058dd1bdc0d42cd1e48431d88c

ParaskP7 committed:
Revert "Database: Update latest database schema for both wp/wc android databases"

This reverts commit 09d1a86b1b5310c6b4d1a9a5cfb3cc6b2dfe90da.

Context: https://github.com/wordpress-mobile/WordPress-FluxC-Android/
pull/3059#discussion_r1688268204

wzieba commented on build.gradle:
```diff
@@ -57,6 +59,12 @@ allprojects {
         debug = false
     }
+
+    tasks.withType(KotlinCompile).all {
```
```suggestion
    tasks.withType(KotlinCompile).configureEach {
```

WDYT about configuring this task lazily? The impact here will be not noticeable, but I think it's a good hygiene to use more performant alternative.

ParaskP7 replied:
This is now done: da917d92a406d669f657acc95f6d413dbee57c84

wzieba committed:
Do not alter context in `CoroutineEngine#flowWithDefaultContext`

As the method doesn't communicate changes to context. Use  `EmptyCoroutineContext` which is default context for `TestCoroutineScope`, but doesn't contain job

wzieba commented on fluxc/src/main/java/org/wordpress/android/fluxc/tools/CoroutineEngine.kt:
```diff
@@ -44,12 +44,14 @@ open class CoroutineEngine
-                    .onStart { appLog.d(tag, "${caller.javaClass.simpleName}: $loggedMessage Started") }
-                    .onEach { appLog.d(tag, "${caller.javaClass.simpleName}: $loggedMessage OnEvent: $it") }
-                    .onCompletion { appLog.d(tag, "${caller.javaClass.simpleName}: $loggedMessage Completed") }
+    ): Flow<RESULT_TYPE> {
+        val safeContext = context.minusKey(Job) // Remove Job from context to make it safe for flows
+        return flow { block() }
+            .flowOn(safeContext)
+            .onStart { appLog.d(tag, "${caller.javaClass.simpleName}: $loggedMessage Started") }
+            .onEach { appLog.d(tag, "${caller.javaClass.simpleName}: $loggedMessage OnEvent: $it") }
+            .onCompletion { appLog.d(tag, "${caller.javaClass.simpleName}: $loggedMessage Completed") }
+    }
```
I've prepared a suggestion in #3062, let's discuss it there

ParaskP7 replied:
Thanks for that @wzieba , much appreciated, let me check on it and we shall proceed accordingly with it. 🙇

ParaskP7 replied:
This is now done: 77e2dfb156dc0025642db0796fd7d0289d02c020

ParaskP7 committed:
Build: Change task with type config from all to configure each

Warning Message: "Consider using 'configureEach' to avoid unnecessary
configuration"

Context: https://github.com/wordpress-mobile/WordPress-FluxC-Android/
pull/3059#discussion_r1689404323

ParaskP7 committed:
Merge pull request #3062 from wordpress-mobile/build/do_not_edit_context

Do not alter context in `CoroutineEngine#flowWithDefaultContext`

ParaskP7 committed:
Merge branch 'build/update-kotlin-to-1.9.25' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/update-kotlin-to-1.9.25

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/update-kotlin-to-1.9.25

wzieba reviewed (approved):
Looks good, great work 🙌 !

ParaskP7 committed:
Deps: Update catalog to 1.18.0 (kotlin 1.9.25)

Catalog: https://github.com/Automattic/android-dependency-catalog/
releases/tag/1.18.0
